sqlmesh/utils/file.py

```diff
@@ -0,0 +1,129 @@
+        """
+        # TODO: pass in a filesystem, naturally
+        self._fs: fsspec.AbstractFileSystem = fsspec.filesystem("file")
+        self._path = path
+        self._original_contents: t.Optional[bytes] = None
+        self._is_locked = False
+
+    @property
+    def path(self) -> str:
+        """The key which represents a path to a file."""
+        return self._path
+
+    @property
+    def lock_path(self) -> str:
+        """The key which represents a path to the file's lock."""
+        return f"{self.path}.lock"
+
+    def read(self) -> t.Optional[bytes]:
+        """Reads data from the file."""
+        if self._fs.exists(self.path):
+            if not self._is_locked:
+                ... # TODO: do whatever we think is right here
+            content = self._fs.cat(self.path)
+            self._original_contents = content[:]
+            return content
+        return None
+    
+    def write(self, content: bytes) -> None:
+        """Writes data within the transaction."""
+        if not self._is_locked:
+            # TODO: do whatever we think is right here
+            raise RuntimeError("Cannot write to a file without a lock.")
+        self._fs.pipe(self.path, content)
+
+    def rollback(self) -> None:
+        """Rolls back the transaction."""
+        if self._original_contents is not None:
+            self._fs.pipe(self.path, self._original_contents)
+
+    def acquire_lock(
+        self, blocking: bool = True, timeout: float = -1
+    ) -> bool:
+        """Acquires a lock on a key.
+        
+        Acquire a lock, blocking or non-blocking.
+
+        When invoked with the blocking argument set to True (the default), block until
+        the lock is unlocked, then set it to locked and return True.
+
+        When invoked with the blocking argument set to False, do not block. If a call
+        with blocking set to True would block, return False immediately; otherwise, set
+        the lock to locked and return True.
+
+        When invoked with the floating-point timeout argument set to a positive value,
+        block for at most the number of seconds specified by timeout and as long as the
+        lock cannot be acquired. A timeout argument of -1 specifies an unbounded wait.
+        If blocking is False, timeout is ignored. The stdlib would raise a ValueError.
+
+        The return value is True if the lock is acquired successfully, False if
+        not (for example if the timeout expired).
+        """
+        if self._is_locked:
+            # Make this idempotent for a single instance.
+            return True
+        lock_id = uuid4()
+        wait_time = 0
+        while self._fs.exists(self.lock_path):
+            if not blocking:
+                return False
+            time.sleep(1)
+            wait_time += 1
+            if timeout > 0 and wait_time > timeout:
+                return False
+        self._fs.pipe(self.lock_path, json.dumps(
```

Comment:
maybe let's just write the str, no need to use json, just the lock-id is fine

Reply:
ok, the reason you'd use json is to store a timeout, if the process crashes with a lock held, you're going to dead lock

so we can keep the id, but we also need to add a lock expire at, that way you can check if a lock is over due and can force it out

Reply:
Yep exactly 😆 I will add

Reply:
Any preference on a default TTL for a lock @tobymao 
You are more familiar with how long a file mutating transaction might realistically be happening

Reply:
most operations should take under 1 second. 

i think maybe the default ttl can be 60 seconds